README.md
# jours-feries

![📆⛱️](https://github.com/SocialGouv/jours-feries/workflows/%F0%9F%93%86%E2%9B%B1%EF%B8%8F/badge.svg) [![codecov](https://codecov.io/gh/SocialGouv/jours-feries/branch/master/graph/badge.svg)](https://codecov.io/gh/SocialGouv/jours-feries)  
[![NPM](https://nodei.co/npm/@socialgouv/jours-feries.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/@socialgouv/jours-feries)

JavaScript / npm port of [jours-feries-france](https://github.com/AntoineAugusti/jours-feries-france) by @AntoineAugusti

## Usage

```js
import joursFeries from "@socialgouv/jours-feries";

console.log(joursFeries(2018));

// include jours feries alsace-moselle
console.log(joursFeries(2018, { alsace: true }));
```

⚠️ Les dates renvoyées sont de vrais objets JavaScript `Date`

```json
{
  "Jour de l'an": "2017-12-31T23:00:00.000Z",
  "Fête du travail": "2018-04-30T22:00:00.000Z",
  "Victoire des alliés": "2018-05-07T22:00:00.000Z",
  "Fête Nationale": "2018-07-13T22:00:00.000Z",
  "Assomption": "2018-08-14T22:00:00.000Z",
  "Toussaint": "2018-10-31T23:00:00.000Z",
  "Armistice": "2018-11-10T23:00:00.000Z",
  "Noël": "2018-12-24T23:00:00.000Z",
  "Lundi de Pâques": "2018-04-01T22:00:00.000Z",
  "Ascension": "2018-05-09T22:00:00.000Z",
  "Lundi de Pentecôte": "2018-05-20T22:00:00.000Z",
  "Vendredi Saint": "2018-03-29T22:00:00.000Z",
  "Saint Étienne": "2018-12-25T23:00:00.000Z"
}
```


CHANGELOG.md
## [3.0.1](https://github.com/SocialGouv/jours-feries/compare/v3.0.0...v3.0.1) (2022-01-13)


### Bug Fixes

* jest in ci (wip) ([5b879ef](https://github.com/SocialGouv/jours-feries/commit/5b879ef2300ee6913f4eb238f86a34c3811c9ec6))

# [3.0.0](https://github.com/SocialGouv/jours-feries/compare/v2.0.0...v3.0.0) (2022-01-13)


* feat!: refactor in TS (#256) ([17a6835](https://github.com/SocialGouv/jours-feries/commit/17a6835ff8aaee3ad78e5d6b70582d8f12585da0)), closes [#256](https://github.com/SocialGouv/jours-feries/issues/256)


### BREAKING CHANGES

* Rename in joursFeries as a named import.
* Refactor of the second parameter, with a zone property.

* fix: Update package.json

Co-authored-by: Julien Bouquillon <contact@revolunet.com>

* update yarn.lock

Co-authored-by: Julien Bouquillon <contact@revolunet.com>

# [2.0.0](https://github.com/SocialGouv/jours-feries/compare/v1.0.1...v2.0.0) (2020-05-11)


### Bug Fixes

* **release:** update package.json file too ([ecae57a](https://github.com/SocialGouv/jours-feries/commit/ecae57a3f920d79e097b756e831e8b42a4bf358c))


* feat(release)!: trigger major release ([e3b3648](https://github.com/SocialGouv/jours-feries/commit/e3b36480afdf10d1498eb24c9f068fdccdae24ea))


### BREAKING CHANGES

* trigger major release

## [1.0.1](https://github.com/SocialGouv/jours-feries/compare/v1.0.0...v1.0.1) (2020-05-11)


### Bug Fixes

* **release:** trigger path release ([ee3f918](https://github.com/SocialGouv/jours-feries/commit/ee3f91841f4b47ab56eced6194c23be270b3ade6))


tests/tests.spec.ts
import { joursFeries } from "../src/index"

test("should match 2018 results", () =>
  expect(joursFeries(2018)).toMatchInlineSnapshot(`
    Object {
      "Armistice": 2018-11-10T23:00:00.000Z,
      "Ascension": 2018-05-09T22:00:00.000Z,
      "Assomption": 2018-08-14T22:00:00.000Z,
      "Fête Nationale": 2018-07-13T22:00:00.000Z,
      "Fête du travail": 2018-04-30T22:00:00.000Z,
      "Jour de l'an": 2017-12-31T23:00:00.000Z,
      "Lundi de Pentecôte": 2018-05-20T22:00:00.000Z,
      "Lundi de Pâques": 2018-04-01T22:00:00.000Z,
      "Noël": 2018-12-24T23:00:00.000Z,
      "Toussaint": 2018-10-31T23:00:00.000Z,
      "Victoire des alliés": 2018-05-07T22:00:00.000Z,
    }
  `))

test("should match 2020 results", () =>
  expect(joursFeries(2020)).toMatchInlineSnapshot(`
    Object {
      "Armistice": 2020-11-10T23:00:00.000Z,
      "Ascension": 2020-05-20T22:00:00.000Z,
      "Assomption": 2020-08-14T22:00:00.000Z,
      "Fête Nationale": 2020-07-13T22:00:00.000Z,
      "Fête du travail": 2020-04-30T22:00:00.000Z,
      "Jour de l'an": 2019-12-31T23:00:00.000Z,
      "Lundi de Pentecôte": 2020-05-31T22:00:00.000Z,
      "Lundi de Pâques": 2020-04-12T22:00:00.000Z,
      "Noël": 2020-12-24T23:00:00.000Z,
      "Toussaint": 2020-10-31T23:00:00.000Z,
      "Victoire des alliés": 2020-05-07T22:00:00.000Z,
    }
  `))

test("should match 2018 alsace-moselle results", () =>
  expect(joursFeries(2018, { zone: "alsace-moselle" })).toMatchInlineSnapshot(`
    Object {
      "Armistice": 2018-11-10T23:00:00.000Z,
      "Ascension": 2018-05-09T22:00:00.000Z,
      "Assomption": 2018-08-14T22:00:00.000Z,
      "Fête Nationale": 2018-07-13T22:00:00.000Z,
      "Fête du travail": 2018-04-30T22:00:00.000Z,
      "Jour de l'an": 2017-12-31T23:00:00.000Z,
      "Lundi de Pentecôte": 2018-05-20T22:00:00.000Z,
      "Lundi de Pâques": 2018-04-01T22:00:00.000Z,
      "Noël": 2018-12-24T23:00:00.000Z,
      "Saint Étienne": 2018-12-25T23:00:00.000Z,
      "Toussaint": 2018-10-31T23:00:00.000Z,
      "Vendredi Saint": 2018-03-29T22:00:00.000Z,
      "Victoire des alliés": 2018-05-07T22:00:00.000Z,
    }
  `))

test("should match 2020 alsace-moselle  results", () =>
  expect(joursFeries(2020, { zone: "alsace-moselle" })).toMatchInlineSnapshot(`
    Object {
      "Armistice": 2020-11-10T23:00:00.000Z,
      "Ascension": 2020-05-20T22:00:00.000Z,
      "Assomption": 2020-08-14T22:00:00.000Z,
      "Fête Nationale": 2020-07-13T22:00:00.000Z,
      "Fête du travail": 2020-04-30T22:00:00.000Z,
      "Jour de l'an": 2019-12-31T23:00:00.000Z,
      "Lundi de Pentecôte": 2020-05-31T22:00:00.000Z,
      "Lundi de Pâques": 2020-04-12T22:00:00.000Z,
      "Noël": 2020-12-24T23:00:00.000Z,
      "Saint Étienne": 2020-12-25T23:00:00.000Z,
      "Toussaint": 2020-10-31T23:00:00.000Z,
      "Vendredi Saint": 2020-04-09T22:00:00.000Z,
      "Victoire des alliés": 2020-05-07T22:00:00.000Z,
    }
  `))

test("Lundi de Pâques 1954", () => {
  expect(joursFeries(1954)["Lundi de Pâques"]).toMatchInlineSnapshot(`1954-04-18T23:00:00.000Z`)
})

test("Lundi de Pâques 1981", () => {
  expect(joursFeries(1981)["Lundi de Pâques"]).toMatchInlineSnapshot(`1981-04-19T22:00:00.000Z`)
})

test("Lundi de Pâques 2049", () => {
  expect(joursFeries(2049)["Lundi de Pâques"]).toMatchInlineSnapshot(`2049-04-18T22:00:00.000Z`)
})


src/__tests__/tests.spec.js
const joursFeries = require("../index");

test("should match 2018 results", () =>
  expect(joursFeries(2018)).toMatchSnapshot());

test("should match 2020 results", () =>
  expect(joursFeries(2020)).toMatchSnapshot());

test("should match 2018 alsace-moselle results", () =>
  expect(joursFeries(2018, { alsace: true })).toMatchSnapshot());

test("should match 2020 alsace-moselle  results", () =>
  expect(joursFeries(2020, { alsace: true })).toMatchSnapshot());

test("Lundi de Pâques 1954", () => {
  expect(joursFeries(1954)["Lundi de Pâques"]).toMatchSnapshot();
});

test("Lundi de Pâques 1981", () => {
  expect(joursFeries(1981)["Lundi de Pâques"]).toMatchSnapshot();
});

test("Lundi de Pâques 2049", () => {
  expect(joursFeries(2049)["Lundi de Pâques"]).toMatchSnapshot();
});


src/index.js
"use strict";

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      configurable: true,
      enumerable: true,
      value: value,
      writable: true,
    });
  } else {
    obj[key] = value;
  }
  return obj;
}

var addDays = function addDays(date, days) {
  return new Date(date.getTime() + days * 24 * 60 * 60 * 1000);
};

var lundiDePaques = function lundiDePaques(year) {
  return addDays(paques(year), 1);
};

var vendrediSaint = function vendrediSaint(year) {
  return addDays(paques(year), -2);
};

var ascension = function ascension(year) {
  return addDays(paques(year), 39);
};

var lundiDePentecote = function lundiDePentecote(year) {
  return addDays(paques(year), 50);
};

var paques = function paques(year) {
  var a = year % 19,
    // century
    b = parseInt(year / 100),
    // years after century
    c = year % 100,
    d =
      (19 * a +
        b -
        parseInt(b / 4) -
        parseInt((parseInt(b - (b + 8) / 25) + 1) / 3) +
        15) %
      30,
    e = (32 + 2 * (b % 4) + 2 * parseInt(c / 4) - d - (c % 4)) % 7,
    f = d + e - 7 * parseInt((a + 11 * d + 22 * e) / 451) + 114,
    month = parseInt(f / 31),
    day = (f % 31) + 1;
  return new Date(year, month - 1, day);
};

var jourDeLAn = function jourDeLAn(year) {
  return new Date(year, 0, 1);
};

var feteDuTravail = function feteDuTravail(year) {
  return new Date(year, 4, 1);
};

var victoireDesAllies = function victoireDesAllies(year) {
  return new Date(year, 4, 8);
};

var feteNationale = function feteNationale(year) {
  return new Date(year, 6, 14);
};

var assomption = function assomption(year) {
  return new Date(year, 7, 15);
};

var toussaint = function toussaint(year) {
  return new Date(year, 10, 1);
};

var armistice = function armistice(year) {
  return new Date(year, 10, 11);
};

var noel = function noel(year) {
  return new Date(year, 11, 25);
};

var saintEtienne = function saintEtienne(year) {
  return new Date(year, 11, 26);
};

var fetes = {
  Armistice: armistice,
  Ascension: ascension,
  Assomption: assomption,
  "Fête Nationale": feteNationale,
  "Fête du travail": feteDuTravail,
  "Jour de l'an": jourDeLAn,
  "Lundi de Pentecôte": lundiDePentecote,
  "Lundi de Pâques": lundiDePaques,
  Noël: noel,
  "Saint Étienne": saintEtienne,
  Toussaint: toussaint,
  "Vendredi Saint": vendrediSaint,
  "Victoire des alliés": victoireDesAllies,
};
var fetesAlsace = ["Vendredi Saint", "Saint Étienne"];

var getJoursFeries = function getJoursFeries(year) {
  var _ref =
      arguments.length > 1 && arguments[1] !== undefined
        ? arguments[1]
        : {
            alsace: false,
          },
    alsace = _ref.alsace;

  return Object.keys(fetes) // filter out alsace if needed
    .filter(function (fete) {
      return alsace || !fetesAlsace.includes(fete);
    }) // get results
    .map(function (fete) {
      return _defineProperty({}, fete, fetes[fete](parseInt(year)));
    }) // flatten dict
    .reduce(function (acc, fete) {
      return Object.assign(acc, fete);
    }, {});
}; //

module.exports = getJoursFeries;


src/index.ts
type ReturnTypeFetesMetropole = ReturnType<typeof fetes>
type ReturnTypeFetesAlsace = ReturnTypeFetesMetropole & ReturnType<typeof fetesAlsace>
type Zones = "métropole" | "alsace-moselle"

const addDays = (date: Date, days: number) => new Date(date.getTime() + days * 24 * 60 * 60 * 1000)

const paques = function (year: number) {
  const a = year % 19
  const century = Math.floor(year / 100)
  const yearsAfterCentury = year % 100
  const d =
    (19 * a + century - Math.floor(century / 4) - Math.floor((Math.floor(century - (century + 8) / 25) + 1) / 3) + 15) %
    30
  const e = (32 + 2 * (century % 4) + 2 * Math.floor(yearsAfterCentury / 4) - d - (yearsAfterCentury % 4)) % 7
  const f = d + e - 7 * Math.floor((a + 11 * d + 22 * e) / 451) + 114
  const month = Math.floor(f / 31)
  const day = (f % 31) + 1

  return new Date(year, month - 1, day)
}

const fetes = (year: number) => ({
  Armistice: new Date(year, 10, 11),
  Ascension: addDays(paques(year), 39),
  Assomption: new Date(year, 7, 15),
  "Fête Nationale": new Date(year, 6, 14),
  "Fête du travail": new Date(year, 4, 1),
  "Jour de l'an": new Date(year, 0, 1),
  "Lundi de Pentecôte": addDays(paques(year), 50),
  "Lundi de Pâques": addDays(paques(year), 1),
  Noël: new Date(year, 11, 25),
  Toussaint: new Date(year, 10, 1),
  "Victoire des alliés": new Date(year, 4, 8),
})

const fetesAlsace = (year: number) => ({
  "Saint Étienne": new Date(year, 11, 26),
  "Vendredi Saint": addDays(paques(year), -2),
})

/**
 * Get french bank holidays by year and zone.
 *
 * @param year year you're interested in
 * @param options the zone you're interested in ("métropole" by default)
 */
function joursFeries(year: number): ReturnTypeFetesMetropole
function joursFeries(year: number, options: { zone: "métropole" }): ReturnTypeFetesMetropole
function joursFeries(year: number, options: { zone: "alsace-moselle" }): ReturnTypeFetesAlsace
function joursFeries(year: number, options: { zone: Zones } = { zone: "métropole" }) {
  if (options.zone === "alsace-moselle") {
    return { ...fetes(year), ...fetesAlsace(year) }
  } else {
    return fetes(year)
  }
}

export { joursFeries }
